src/radical/pilot/raptor/worker_mpi.py

```diff
@@ -554,47 +612,68 @@ def _dispatch_function(self, task, env):
         func = task['description']['function']
 
         to_call = ''
+        names   = ''
         args    = task['description'].get('args',   [])
         kwargs  = task['description'].get('kwargs', {})
+        py_func = False
+
+        self._log.debug('=== orig args: %s : %s', args, kwargs)
 
         # check if we have a serialized object
+        self._log.debug('func serialized: %d: %s', len(func), func)
         try:
-
+            # FIXME: can we have a better test than try/except?  This hides
+            #        potential errors...
+            # FIXME: ensure we did not get args and kwargs from above
             to_call, args, kwargs = PythonTask.get_func_attr(func)
-
-            # Inject the communicator in the kwargs
-            if task['description'].get('cpu_process_type') == RP_MPI:
-                kwargs['comm'] = task['description']['args'][0]
-
-
-        except Exception:
-            self._log.error('failed to obtain callable from task function')
+            py_func = True
+        except:
+            pass
 
         if not to_call:
             assert(func)
             # check if `func_name` is a global name
             names   = dict(list(globals().items()) + list(locals().items()))
             to_call = names.get(func)
 
-
         # if not, check if this is a class method of this worker implementation
         if not to_call:
             to_call = getattr(self._base, func, None)
 
-
         if not to_call:
             self._log.error('no %s in \n%s\n\n%s', func, names, dir(self._base))
             raise ValueError('callable %s not found: %s' % (to_call, task))
 
+        comm = task.get('mpi_comm')
+        if comm:
+            # we have an MPI communicator we need to inject into the function's
+            # arguments.
+            if py_func:
+                # For a `py_func` we add the communicator as `comm` kwarg if
+                # that is set to None, and otherwise as first `arg` if that is
+                # None.  If neither is true we'll error out.
+                # NOTE that we don't change the number of arguments either way.
+                if 'comm' in kwargs and kwargs['comm'] is None:
```

Comment:
TODO AM: make sure this still works

Reply:
This does work, it was a silly mistake from my side, sorry for the noise.